### PR TITLE
Scan for resolvers that have a supertype of the data class

### LINKED ISSUE #497: FieldResolverError with resolver for superclass or interface
## Description

Consider the following dataclass hierarchy:
```java
abstract class Base {
}

abstract class Thing extends Base {
}
```

This is a custom resolver for the abstract class `Base`:
```java
class BaseResolver implements GraphQLResolver<Base> {
    public boolean isActive(Base base) {
       return true;
   }
}
```

Root query resolver with datafetcher that returns our concrete class `Thing`:
```java
class RootQueryResolver implements GraphQLQueryResolver {
    public Thing getThing() {
        return new Thing();
    }
}
```

### Expected behavior
The following schema should work fine and return true for queries on the `active` field:
```graphqls
type Query {
    thing: Thing!
}

type Thing {
    active: Boolean!
}
```

### Actual behavior
```
graphql.kickstart.tools.resolver.FieldResolverError: No method or field found as defined in schema [...]
```

### Steps to reproduce the bug

1. Create a resolver for a supertype (class or interface) with any datafetcher
2. Have a query return a subclass of that type
3. Try build the schema

### PR DESCRIPTION
**Pull request by florensie**
Fixes #497

## Checklist
- [x] Pull requests follows the [contribution guide](https://github.com/graphql-java-kickstart/graphql-java-tools/wiki/Contribution-guide)
- [x] New or modified functionality is covered by tests

## Description
Made necessary changes to class scanning logic to support using supertypes in resolvers.

### CODE REVIEW

File: src/main/kotlin/graphql/kickstart/tools/SchemaClassScanner.kt [hunk 1]
```diff
@@ -249,14 +249,18 @@ internal class SchemaClassScanner(
...
+        val resolverInfoList = this.resolverInfos.filter { it.dataClassType.unwrap().isAssignableFrom(item.clazz.unwrap()) }
+        val resolverInfo: ResolverInfo = (if (resolverInfoList.size > 1 && item.clazz != Object::class.java) {
+            MultiResolverInfo(resolverInfoList, item.clazz.unwrap())
         } else {
             if (item.clazz == Object::class.java) {
                 getResolverInfoFromTypeDictionary(item.type.name)
             } else {
-                resolverInfosByDataClass[item.clazz] ?: DataClassResolverInfo(item.clazz)
+                if (resolverInfoList.size == 1) {
+                    MultiResolverInfo(resolverInfoList, item.clazz.unwrap())
```
Comment (oryan-block) [inline src/main/kotlin/graphql/kickstart/tools/SchemaClassScanner.kt:260]: Why multi if there's only one?

File: src/main/kotlin/graphql/kickstart/tools/resolver/FieldResolverScanner.kt [hunk 2]
```diff
@@ -127,7 +127,8 @@ internal class FieldResolverScanner(val options: SchemaParserOptions) {
     private fun verifyMethodArguments(method: Method, requiredCount: Int, search: Search): Boolean {
         val appropriateFirstParameter = if (search.requiredFirstParameterType != null) {
             method.genericParameterTypes.firstOrNull()?.let {
-                it == search.requiredFirstParameterType || method.declaringClass.typeParameters.contains(it)
+                it == search.requiredFirstParameterType || (it is Class<*> && it.unwrap().isAssignableFrom(search.requiredFirstParameterType))
```
Comment (oryan-block) [inline src/main/kotlin/graphql/kickstart/tools/resolver/FieldResolverScanner.kt:130]: Do we need the cast to Class here?

**Review by oryan-block — CHANGES_REQUESTED** [review]
Thanks for the contribution 👍 
I have some questions.

**Comment by swollner** [comment]
Hi is there any progress on this?

**Comment by florensie** [comment]
@swollner None beyond this PR. Best to keep discussion to #497, feel free to use this PR as a reference/base if you wish to contribute yourself.

**Pull request closed without merging**